source/common/http/conn_manager_impl.cc

```diff
@@ -1192,6 +1192,9 @@ void ConnectionManagerImpl::ActiveStream::refreshDurationTimeout() {
       const auto max_stream_duration = connection_manager_.config_.maxStreamDuration();
       if (max_stream_duration.has_value() && max_stream_duration.value().count()) {
         timeout = max_stream_duration.value();
+      } else if (route->timeout().count() != 0) {
+        // If max stream duration is not set either at route/HCM level, use the route timeout.
+        timeout = route->timeout();
```

Comment:
Yeah, I think think this is right.
I'm going to tag @antoniovicente for second pass of code as a second eye to make sure we got the various combinations and permutations right, and I'll do a second pass when you've got the tests sorted out.

Reply:
I'm wondering if something else in the system is already enforcing the relevant timeout.  I found use of timeout_.global_timeout_ in Filter::onRequestComplete, but I think that timer is not comprehensive as it waits until the full request body has been received before it is activated.  I wonder if we could use the same logic to compute the timeout in both places and/or have a single timer to control this behavior.

https://github.com/envoyproxy/envoy/blob/0314715ee7cfaa72b57aa592b13d89f4b90cb9c1/source/common/router/router.cc#L830

Reply:
AFAIK it is difficult to unify that logic as that one is meant with old timeouts. @alyssawilk any thoughts?

Reply:
Yeah, I think unifying is outside the scope of this PR so I'd be inclined to just land the simpler fix.

Reply:
Ok.  I think that the timeout handling added in this PR is fine.